src/connect.rs

```diff
@@ -28,10 +29,80 @@ use self::native_tls_conn::NativeTlsConn;
 #[cfg(feature = "rustls-tls")]
 use self::rustls_tls_conn::RustlsTlsConn;
 
-#[cfg(feature = "trust-dns")]
-type HttpConnector = hyper::client::HttpConnector<TrustDnsResolver>;
-#[cfg(not(feature = "trust-dns"))]
-type HttpConnector = hyper::client::HttpConnector;
+#[derive(Clone)]
+pub(crate) enum HttpConnector {
+    Gai(hyper::client::HttpConnector),
+    #[cfg(feature = "trust-dns")]
+    TrustDns(hyper::client::HttpConnector<TrustDnsResolver>),
+}
+
+impl HttpConnector {
+    pub(crate) fn new_gai() -> Self {
+        Self::Gai(hyper::client::HttpConnector::new())
+    }
+
+    #[cfg(feature = "trust-dns")]
+    pub(crate) fn new_trust_dns() -> crate::Result<HttpConnector> {
+        TrustDnsResolver::new()
+            .map(hyper::client::HttpConnector::new_with_resolver)
+            .map(Self::TrustDns)
+            .map_err(crate::error::builder)
+    }
+}
+
+macro_rules! impl_http_connector {
+    ($(fn $name:ident(&mut self, $($par_name:ident: $par_type:ty),*)$( -> $return:ty)?;)+) => {
+        #[allow(dead_code)]
+        impl HttpConnector {
+            $(
+                fn $name(&mut self, $($par_name: $par_type),*)$( -> $return)? {
+                    match self {
+                        Self::Gai(resolver) => resolver.$name($($par_name),*),
+                        #[cfg(feature = "trust-dns")]
+                        Self::TrustDns(resolver) => resolver.$name($($par_name),*),
+                    }
+                }
+            )+
+        }
+    };
+}
+
+impl_http_connector! {
+    fn set_local_address(&mut self, addr: Option<IpAddr>);
+    fn enforce_http(&mut self, is_enforced: bool);
+    fn set_nodelay(&mut self, nodelay: bool);
+}
+
+impl Service<Uri> for HttpConnector {
+    type Response = <hyper::client::HttpConnector as Service<Uri>>::Response;
+    type Error = <hyper::client::HttpConnector as Service<Uri>>::Error;
+    #[cfg(feature = "trust-dns")]
+    type Future = Either<
+        <hyper::client::HttpConnector as Service<Uri>>::Future,
+        <hyper::client::HttpConnector<TrustDnsResolver> as Service<Uri>>::Future,
+    >;
+    #[cfg(not(feature = "trust-dns"))]
+    type Future = Either<
+        <hyper::client::HttpConnector as Service<Uri>>::Future,
+        <hyper::client::HttpConnector as Service<Uri>>::Future,
+    >;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        match self {
+            Self::Gai(resolver) => resolver.poll_ready(cx),
+            #[cfg(feature = "trust-dns")]
+            Self::TrustDns(resolver) => resolver.poll_ready(cx),
+        }
+    }
+
+    fn call(&mut self, dst: Uri) -> Self::Future {
+        match self {
+            Self::Gai(resolver) => Either::Left(resolver.call(dst)),
+            #[cfg(feature = "trust-dns")]
+            Self::TrustDns(resolver) => Either::Right(resolver.call(dst)),
+        }
+    }
+}
```

Comment:
I could instead do a [`impl AsMut<tower_util::Either> for HttpConnector`](https://docs.rs/tower-util/0.3.0/tower_util/enum.Either.html)?